Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1

```diff
@@ -20,30 +20,99 @@ Write-Verbose "cleanTargetBeforeCopy = $cleanTargetBeforeCopy" -Verbose
 
 import-module "Microsoft.TeamFoundation.DistributedTask.Task.DevTestLabs"
 
-$connection = Get-VssConnection -TaskContext $distributedTaskContext
+	# Default + constants #
+$defaultWirmPort = '5985'
+$defaultSkipCACheckOption = '-SkipCACheck'
+$defaulthttpProtocallOption = '-UseHttp'
+$resourceFQDNKeyName = 'Microsoft-Vslabs-MG-Resource-FQDN'
+$resourceWinRMHttpPortKeyName = 'WinRM_HttpPort'
+$resourceWinRMHttpsPortKeyName = 'WinRM_HttpsPort'
+$envOperationStatus = 'Passed'
+
+function Get-ResourceCredentials
+{
+	param([object]$resource)
+		
+	$machineUserName = $resource.Username
+	Write-Verbose "`t`t Resource Username - $machineUserName" -Verbose
+	$machinePassword = $resource.Password
 
-$resources = Get-EnvironmentResources -EnvironmentName $environmentName -ResourceFilter $machineNames -Connection $connection -ErrorAction Stop
+	$credential = New-Object 'System.Net.NetworkCredential' -ArgumentList $machineUserName, $machinePassword
+	
+	return $credential
+}
+
+function Get-ResourcesProperties
+{
+    param([object]$resources)
 
-$machineUserName = Get-EnvironmentProperty -EnvironmentName $environmentName -Key "Username" -Connection $connection -ErrorAction Stop
+	$resourcesProperties = @()
+	
+	foreach ($resource in $resources)
+    {
+		[hashtable]$eachResourceProperties = @{} 
+		
+		$resourceName = $resource.Name
+		Write-Verbose "Get Resource properties for $resourceName" -Verbose		
+		
+		$fqdn = Get-EnvironmentProperty -EnvironmentName $environmentName -Key $resourceFQDNKeyName -Connection $connection -ResourceName $resourceName -ErrorAction Stop
+		
+		Write-Verbose "`t`t Resource fqdn - $fqdn" -Verbose	
+		
+		$eachResourceProperties.fqdn = $fqdn
 
-$machinePassword = Get-EnvironmentProperty -EnvironmentName $environmentName -Key "Password" -Connection $connection -ErrorAction Stop
+		$eachResourceProperties.httpProtocallOption = ''
+		$eachResourceProperties.skipCACheckOption = $defaultSkipCACheckOption
+
+		$winrmPort = Get-EnvironmentProperty -EnvironmentName $environmentName -Key $resourceWinRMHttpsPortKeyName -Connection $connection -ResourceName $resourceName -ErrorAction Stop
+		
+		if([string]::IsNullOrEmpty($winrmPort))
+		{
+			$winrmPort = Get-EnvironmentProperty -EnvironmentName $environmentName -Key $resourceWinRMHttpPortKeyName -Connection $connection -ResourceName $resourceName -ErrorAction Stop
+			if([string]::IsNullOrEmpty($winrmPort))
+			{
+				Write-Verbose "`t`t Resource $resourceName does not have any winrm port defined , use the default - $defaultWirmPort" -Verbose
+				$winrmPort = $defaultWirmPort
+			}
+			$eachResourceProperties.httpProtocallOption = $defaulthttpProtocallOption
+		}
+		else
+		{
+			# Do not use 'SkipCACheck' option in case of winrm https port
+			$eachResourceProperties.skipCACheckOption = ''
+			Write-Verbose "`t`t Resource $resourceName has winrm https port $winrmPort defined , CA option will be used to expand the target path if it is provided as environment variable " -Verbose
```

Comment:
Even if user provides this as environment variable, where are you fetching it? I don't see Get-EnvironmentProperty for skipCA option and validating it? I thought SkipCA will be exposed in Task UI and not taken as environment variable. For CopyFiles, this should be hidden under Advanced group as for most cases we are good with http.